src/effgen.ml

```diff
@@ -600,64 +600,53 @@ module GeneratorsWithContext (Ctx : Context) = struct
       [ (3, gen) ]
     | _ -> []
 
-  (* [gen_term_from_rules env goal size rules] returns a constant generator with a term
-     wrapped in an option, in which the term was generated using a randomly picked
-     generator from the {!rules} list
+  (* [term_from_rules rules] returns a constant generator with a term option,
+     in which the term was generated using a randomly picked generator from [rules].
 
      Generators are picked according to their weights.
   *)
-  and gen_term_from_rules _env _goal _size rules =
-    let int_below_gen bound st = Random.State.int st bound in
-    let weights_sum = List.fold_left (fun acc (w, _g) -> acc + w) 0 rules in
-    (* we reimplement QCheck.Gen.frequency because we want it to also return the weight of
-       the picked element (to update weights sum) and the list without that element *)
-    let gen_freq lst rand_k =
-      let rec pick lst (acc_lst, acc_sum) =
-        match lst with
-        | [] -> failwith "gen_term_from_rules: gen_freq: too large rand_k"
-        | ((w, _g) as curr) :: rest ->
-          if rand_k < acc_sum + w
-          then (curr, List.rev_append acc_lst rest)
-          else pick rest (curr :: acc_lst, acc_sum + w)
+  and term_from_rules rules : term option Gen.t =
+   fun st ->
+    (* TODO: use [QCheck.Gen.float_bound_inclusive] and [QCheck.Gen.shuffle_w_l] *)
+    let float_bound_inclusive bound st = Random.State.float st bound in
+    let shuffle_w_l l st =
+      let sample (w, v) =
+        let fl_w = float_of_int w in
+        (float_bound_inclusive 1. st ** (1. /. fl_w), v)
       in
-      pick lst ([], 0)
+      let samples = List.map sample l in
+      List.sort (fun (w1, _) (w2, _) -> compare w2 w1) samples |> List.map snd
```

Comment:
I would move `float_bound_inclusive` and `shuffle_w_l` to auxiliary functions at the beginning of the module, so that the domain-specific code in this function is shorter and easier to read.

Reply:
OK. I thought that as those functions are not used anywhere else, it is best to keep them local. Your idea makes sense.

Reply:
Generally I try to make a distinction between functions that are not used anywhere but "would make sense" somewhere else, and functions (or in general declarations) that only make sense in the local context.